### ISSUE TITLE
Discussion: Should it be easier to target specific Windows versions with conditional XAML?

### ISSUE THREAD

**Issue by Marv51** [microsoft/microsoft-ui-xaml#6546]
## Discussion:  Should it be easier to target specific Windows versions with conditional XAML? 
Look, I know writing code for a specific version of Windows is the root of many problems. But, some features (even in WinUI 3) require at least a certain version of Windows. For example, Segoe Fluent Icons are only available on Windows 11.

In the UWP everywhere (_universal_?) vision of the future, it made sense to version every API contract separately. That way different devices could target different sets of these APIs. Is that really still the case? My app only targets traditional Windows and that will likely stay that way. 

So instead of 
```xaml
<Page
xmlns:contract5Present="http://schemas.microsoft.com/winfx/2006/xaml/presentation?IsApiContractPresent(Windows.Foundation.UniversalApiContract,5)">
```
wouldn't it be much clearer to create something like this:
```xaml
<Page
xmlns:contract5Present="http://schemas.microsoft.com/winfx/2006/xaml/presentation?IsMinimumWindowsVersion(16299)">
```

Windows build numbers are already pretty complicated, but these ApiContract versions are even less familiar to most developers.

### Benefits:
* Cleaner and easier to understand syntax
* Easy to find old conditional xaml that is no longer needed
* No need to translate numbers between various versioning schemes

### Related Links
At a minimum these API Contract Versions need to be much better documented, because we can't always refer to this [one comment](https://github.com/WICG/ua-client-hints/issues/220#issuecomment-870858413) on GitHub to find out what versions of Windows they actually refer to.

### Notes
The inverse method eqivalent to IsApiContractNotPresent is also needed.
The naming of this method might need some work. Alternatives: WindowsVersionIsLowerThen, IsAtLeastBuildNumber

**Comment by mrlacey** [comment]
Does Conditional XAML work in WinUI3? I thought it was only supported in UWP (WinUI2) projects?
If not, then can/should/will it be?

Regardless of whether it does or not, assuming it does (or will) I think the above proposal is too limited.

The [current UWP support for conditional XAML](https://docs.microsoft.com/en-us/windows/uwp/debug-test-perf/conditional-xaml) is very limited.

Instead, I'd like to see it expanded so it takes the form:

```
xmlns:myNamespace="schema?staticClass.conditionalMethod(parameter)"
```

By adding `staticClass` to the directive it could then work with any available static method at runtime. (All the conditional XAML evaluation is runtime only anyway.)
`parameter` could also be made optional.  I do not think it necessary to support multiple parameters or a dynamic value for the parameter.

In terms of design/build time evaluation, it would (I think) be sufficient to only verify that the static class and method exist and take the defined number of parameters.

The problem with the current (UWP) implementation is that it is restricted to only three methods (& their inverses) and only one of them has decent VS support.

Rather than, as the OP requests, add the ability to use a single new method, enable the framework to let us call any [static] method and then the OP can implement the method they want in any way they want. And any other developer can add any other functionality they want too. (Things like the OP suggests could be useful additions to the community toolkit, once the platform makes them possible.)
I can see the ability to be able to use this as a great way to reduce overall code in XAML files, and the need to have logic in the accompanying code-behind file.


I appreciate that any changes to XAML are complex and rarely suggested but this could be a very powerful addition.

**Comment by Marv51** [comment]
@mrlacey Love that idea, good generalization of my suggestion. 

I can confirm that conditional XAML works in WinUI 3/WinAppSDK 1.0.

The namespace of the static class in your example would need to be fully qualified, right?

**Comment by mrlacey** [comment]
> @mrlacey Love that idea, good generalization of my suggestion. 
>
> The namespace of the static class in your example would need to be fully qualified, right?

I was assuming the the namespace of the static class would be in the schema

**Comment by dansmor7** [comment]
I agree, it should at least be better documented. I think it would be pretty easy to provide that table with the versions and contract numbers somewhere officially.
